spock-specs/src/test/groovy/org/spockframework/runtime/StoreSpec.groovy

```diff
@@ -0,0 +1,343 @@
+    runner.addClassImport(FailingStoreUsage)
+    runner.addClassMemberImport(MethodKind)
+  }
+
+  def "store is created and cleaned for each level"() {
+    given:
+    LoggingStoreInterceptor.HACKY_SHARED_LIST_OF_ACTIONS.clear()
+
+    when:
+    runner.runWithImports('''
+    @LogStoreUsage
+    class ASpec extends Specification {
+      def "a feature"() {
+        expect: true
+      }
+
+      def "data driven feature"() {
+        expect: true
+        where:
+        i << [1, 2]
+      }
+    }
+''')
+
+    then:
+    LoggingStoreInterceptor.HACKY_SHARED_LIST_OF_ACTIONS == [
+      'before SHARED_INITIALIZER - null',
+      '	prev:     null',
+      '	replaced: null',
+      '	put:      Stored at SHARED_INITIALIZER - null',
+      'after SHARED_INITIALIZER - null',
+      'before SPEC_EXECUTION - null',
+      '	prev:     Stored at SHARED_INITIALIZER - null',
+      '	replaced: Stored at SHARED_INITIALIZER - null',
+      '	put:      Stored at SPEC_EXECUTION - null',
+      'before SETUP_SPEC - null',
+      '	prev:     Stored at SPEC_EXECUTION - null',
+      '	replaced: Stored at SPEC_EXECUTION - null',
+      '	put:      Stored at SETUP_SPEC - null',
+      'after SETUP_SPEC - null',
+      'before INITIALIZER - null',
+      '	prev:     Stored at SETUP_SPEC - null',
+      '	replaced: null',
+      '	put:      Stored at INITIALIZER - null',
+      'after INITIALIZER - null',
+      'before FEATURE_EXECUTION - null',
+      '	prev:     Stored at INITIALIZER - null',
+      '	replaced: Stored at INITIALIZER - null',
+      '	put:      Stored at FEATURE_EXECUTION - null',
+      'before ITERATION_EXECUTION - null',
+      '	prev:     Stored at FEATURE_EXECUTION - null',
+      '	replaced: Stored at FEATURE_EXECUTION - null',
+      '	put:      Stored at ITERATION_EXECUTION - null',
+      'before SETUP - null',
+      '	prev:     Stored at ITERATION_EXECUTION - null',
+      '	replaced: Stored at ITERATION_EXECUTION - null',
+      '	put:      Stored at SETUP - null',
+      'after SETUP - null',
+      'before FEATURE - a feature',
+      '	prev:     Stored at SETUP - null',
+      '	replaced: Stored at SETUP - null',
+      '	put:      Stored at FEATURE - a feature',
+      'after FEATURE - a feature',
+      'before CLEANUP - null',
+      '	prev:     Stored at FEATURE - a feature',
+      '	replaced: Stored at FEATURE - a feature',
+      '	put:      Stored at CLEANUP - null',
+      'after CLEANUP - null',
+      '# closing Stored at CLEANUP - null',
+      '# closing Stored at FEATURE - a feature',
+      '# closing Stored at SETUP - null',
+      '# closing Stored at ITERATION_EXECUTION - null',
+      '# closing Stored at FEATURE_EXECUTION - null',
+      '# closing Stored at INITIALIZER - null',
+      'after ITERATION_EXECUTION - null',
+      'after FEATURE_EXECUTION - null',
+      'before FEATURE_EXECUTION - null',
+      '	prev:     Stored at SETUP_SPEC - null',
+      '	replaced: Stored at SETUP_SPEC - null',
```

Comment:
Effect of too late child store provider creation mentioned above, should be `null`.